### ISSUE TITLE
NO more ipv6 on all interfaces on Fedora 25 package docker-engine-17.04.0.ce-1.fc25.x86_64

### ISSUE THREAD

**Issue by vgallissot** [docker/docker#32433]
**Description**
NO IPv6 address is available since docker-engine-17.04.0.ce-1.fc25.x86_64, not even on loop.

**Steps to reproduce the issue:**
1. On Fedora 25 x86_64: Update docker-engine to docker-engine-17.04.0.ce-1.fc25.x86_64
2. Docker run some image
3. No IPv6 address is available regardless of the interface

**Describe the results you received:**
Example:
```shell
root@mycont1:/# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
13: eth0@if14: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.2/16 scope global eth0
       valid_lft forever preferred_lft forever
root@mycont1:/#
```
In that example, I used ubuntu:14.04 official docker image.

**Describe the results you expected:**
To have inet6 entries when showing interfaces.  
Package downgrade gives my IPv6 addresses back: ``dnf downgrade docker-engine docker-engine-selinux``

Example:
```shell
docker run -t -i --rm --name "mycont1" -h "mycont1"  "ubuntu:14.04" /bin/bash
root@mycont1:/# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
15: eth0@if16: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.2/16 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe11:2/64 scope link tentative 
       valid_lft forever preferred_lft forever
```

**Additional information you deem important (e.g. issue happens only occasionally):**
Downgrading package to **17.03.1.ce-1.fc25** fixes the issue

**Output of `docker version`:**

```
± docker version
Client:
 Version:      17.04.0-ce
 API version:  1.28
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Mon Apr  3 18:06:13 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.04.0-ce
 API version:  1.28 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Mon Apr  3 18:06:13 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
± docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 17.04.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: 
containerd version: 422e31ce907fd9c3833a38d7b8fdd023e5a76e73
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.10.8-200.fc25.x86_64
Operating System: Fedora 25 (Workstation Edition)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 15.35GiB
Name: vincentg-lap2
ID: 5JJQ:CP3G:POLH:NLLI:WSSP:IENF:QAAA:Q3CJ:QDRZ:6TEJ:3H6O:TL6V
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
nothing that matters

**Comment by cpuguy83** [comment]
ping @aboch

**Comment by aboch** [comment]
It's because of #20569

**Comment by aboch** [comment]
You now need to enable `--ipv6` and provide a `--fixed-cidr-v6` when booting the daemon, if you want the interfaces of containers running on the default `bridge` network to have the ipv6 addresses.

**Comment by adudek16** [comment]
Glad I found this. Just ran into this issue. We just use IPv6 and this was killing the ipv6 on the ovs bridge that we wanted to connect to the containers to.  Question: Is that expected? Should docker clobber settings outside of it?

**Comment by atoz7** [comment]
hello,

i had the same problem with the prepackaged docker-engine debian-package from apt.dockerproject.org -> version 17.04.0~ce-0~debian-jessie

idk ... i don't think its good for the package to *entirely disable ipv6 on the host-machine* - no matter what (ipv6) settings are present for the docker-engine package/daemon...

imho: such a major "side-effect" sure violate some packaging rules for debian ;)

regards,
joachim

**Comment by aboch** [comment]
>  i don't think its good for the package to entirely disable ipv6 on the host-machine

The change does not disable ipv6 on the host machine.
The change disable ipv6 on the container if it is not connected to any IPv6 docker network.

**Comment by atoz7** [comment]
hello,

sorry ... but in my experience on debian with the current docker-engine package:
the ipv6 from the *host* got disabled.

again: the ipv6-networking from the *host* not the *container* ...

regards,
joachim

**Comment by adudek16** [comment]
I can confirm that it is disabling ipv6 globally.
did a sysctl -a | grep disable_ipv6 under 17.03.1 and again under 17.04 and yes docker changed them all to "1"

**Comment by cpuguy83** [comment]
This seems to happen with `--net=host` @aboch

**Comment by aboch** [comment]
Thank you @cpuguy83 . That explains it.
Will take care of it.

**Comment by aboch** [comment]
@adudek16 @atoz7 I am pushing a fix for this now on `--network host`.
But can you please confirm you did run a container with `--network host` setting for the issue to happen. Thanks.

**Comment by atoz7** [comment]
hello,

yes ... i had at least one ctr running with "--network host".

imho. it would be nice, if in the default installation the ipv6 settings wouldn't be touched or even set to enabled.

in 2017 ipv6/v4 dualstack operation should be pretty common by now.

i never took a detailed look at it, but if it needs a network to work ... just use a fe80::* local subnet address, if there is no subnet configured :)

regards,
joachim

**Comment by adudek16** [comment]
@aboch Yes. ctr is running with --net=host

**Comment by vgallissot** [comment]
I couldn't answer before sorry.

Reading backlog and I see relations to --net=host.
I don't do that. **I don't use --net at all.**

That means docker proxy disables ipv6 by default.

To me, problem is still not fixed.

cc @cpuguy83 @aboch @mavenugo

**Comment by cpuguy83** [comment]
@vgallissot Are you able to test the fix?

**Comment by vgallissot** [comment]
not yet, but the PR  #32447 only fixes --network=host which I never use.

Testing asap

**Comment by aboch** [comment]
@vgallissot 
From your description of this issue, you were complaining that the container interfaces no longer have IPv6. This is by design if the container is not connected to a IPv6 enabled network (after #20569).
From now on you need to enable IPv6 on the network where you are running your containers, if you want its interfaces to have the IPv6 addresses. We are not planning to revert this behavior.

Then others added comments to this issue to report an important regression where running a container with net mode `host` would disable IPv6 on the host. This problem took over and your issue in fact became the "host IPv6 removal issue"... therefore we closed it once we fixed the regression.

Probably we should have asked the other reporters to open a new issue for that.

**Comment by atoz7** [comment]
hello,

oops - sorry for that -> didn't want to "squat" your issue, but at first i thought, i am encountering the same problems as @vgallissot and i didn't want to create a duplicate :)

regards,
joachim

**Comment by vgallissot** [comment]
Ok got it, I'll try some workaround.
I'm just estonished on that decision to completely remove IPv6 stack by default.
Regards

**Comment by SpComb** [comment]
Upgrading to 17.04 which disables IPv6 on the container `lo` interface also breaks any services that default to binding on `::1`. This affects e.g. [`nsd`](https://www.nlnetlabs.nl/projects/nsd/):

```
[2017-04-27 07:51:27.952] nsd[11]: error: can't bind tcp socket: Cannot assign requested address
[2017-04-27 07:51:27.952] nsd[11]: error: cannot open control interface ::1 8952
[2017-04-27 07:51:27.952] nsd[11]: error: could not open remote control port
[2017-04-27 07:51:27.952] nsd[11]: error: could not perform remote control setup
```

The same configuration used to work with Docker 1.13 before the upgrade to 17.04.

I can't seem to find any open issues that would address this... I will post a separate issue next, I suppose?

**Comment by chpio** [comment]
> I can't seem to find any open issues that would address this... I will post a separate issue next, I suppose?

that's because it's fixed now... you have just to wait a few years for the update to appear.

**Comment by adudek16** [comment]
It should be in the 17.05 rc1

**Comment by SpComb** [comment]
>> I can't seem to find any open issues that would address this... I will post a separate issue next, I suppose?
> 
> that's because it's fixed now... you have just to wait a few years for the update to appear.

The only fix I can see is  #32447 to fix the `--net=host` issue (#32639). That does not fix the missing `::1` address, which may break containers that used to work in earlier versions of Docker.

**Comment by chpio** [comment]
> The only fix I can see is #32447 to fix the --net=host issue (#32639). That does not fix the missing ::1 address, which may break containers that used to work in earlier versions of Docker.

but the missing `::1` address is the result of disabling IPv6 on all interfaces of the host, isn't it?

**Comment by SpComb** [comment]
>> The only fix I can see is #32447 to fix the --net=host issue (#32639). That does not fix the missing ::1 address, which may break containers that used to work in earlier versions of Docker.
> 
> but the missing `::1` address is the result of disabling IPv6 on all interfaces of the host, isn't it?

Docker 17.04.0 seems to disable IPv6 by default for all containers, including both the normal `--net=bridge` and `--net=host` containers. For normal `--net=bridge` containers, that means that applications will now be unable to bind to `::1` within the containers. For `--net=host` containers, that also has the side-effect of disabling IPv6 within the host's network namespace, but that's a separate issue fixed by #32447.

Disabling the `::1` IPv6 loopback address for the `lo` interface seems to be an intentional part of #20569, apparently as a workaround for the `unregister_netdevice: waiting for lo to become free. Usage count = 3`(#5618) issue? It just introduces a regression for applications such as `nsd`, that have been binding to `::1` as part of their default configuration. Starting the same container with Docker 17.04, and the application will no longer be able to bind to `::1` , which may cause the application to fail, even though it works working before Docker 17.04.

**Comment by adudek16** [comment]
Sounds like a new bug

**Comment by chpio** [comment]
I just tried to run dockerd in IPv6 mode (https://docs.docker.com/engine/userguide/networking/default_network/ipv6/) providing `--ipv6` and `--fixed-cidr-v6`, but it just disabled IPv6. How should it run in IPv6 mode while disabling IPv6? Is the support for IPv6 just some kind of *unstable*/*work-in-progress* (it exists but you shouldn't expect it to work in any meaningful way) in docker?

**Comment by euank** [comment]
@SpComb I agree that's a new bug, so I created a new issue over here: #33099
